cert/cert.go

```diff
@@ -144,47 +181,145 @@ func NewLeaf(commonName, caCertPEM, caKeyPEM string, expiry time.Time, hostnames
 	if err != nil {
 		return "", "", errors.Trace(err)
 	}
+
+	subject := pkix.Name{
+		CommonName:   commonName,
+		Organization: []string{"juju"},
+		SerialNumber: UUID,
+	}
+
+	// get asn1 encoded info of the subject pkix
+	value, err := getUPNExtensionValue(subject)
+	if err != nil {
+		return "", "", fmt.Errorf("Can't marshal asn1 encoded %s", err)
+	}
+
 	// TODO(perrito666) 2016-05-02 lp:1558657
 	now := time.Now()
+
 	template := &x509.Certificate{
+		Subject:      subject,
 		SerialNumber: serialNumber,
-		Subject: pkix.Name{
-			// This won't match host names with dots. The hostname
-			// is hardcoded when connecting to avoid the issue.
-			CommonName:   commonName,
-			Organization: []string{"juju"},
-		},
-		NotBefore: now.UTC().AddDate(0, 0, -7),
-		NotAfter:  expiry.UTC(),
-
+		NotBefore:    now.UTC().AddDate(0, 0, -7),
+		Version:      2,
+		NotAfter:     expiry.UTC(),
 		SubjectKeyId: bigIntHash(key.N),
 		KeyUsage:     x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageKeyAgreement,
 		ExtKeyUsage:  extKeyUsage,
+		ExtraExtensions: []pkix.Extension{
+			{
+				Id:       subjAltName,
+				Critical: false,
+				Value:    value,
+			},
+		},
 	}
+
 	for _, hostname := range hostnames {
 		if ip := net.ParseIP(hostname); ip != nil {
 			template.IPAddresses = append(template.IPAddresses, ip)
 		} else {
 			template.DNSNames = append(template.DNSNames, hostname)
 		}
 	}
-	certDER, err := x509.CreateCertificate(rand.Reader, template, caCert, &key.PublicKey, caKey)
+
+	if caCertPEM == "" &&
+		caKeyPEM == "" {
+		caCert = template
+		caKey = key
+	}
+
+	certDER, err := x509.CreateCertificate(rand.Reader, template, caCert, getPublicKey(key), caKey)
```

Comment:
This is not fixed. Did you miss a commit? Maybe I'm being unclear.

What I want to see is:
 - no call to x509.CreateCertificate in this function
 - calling NewLeaf from github.com/juju/utils/cert (https://github.com/juju/utils/blob/master/cert/cert.go#L119) to create the certificate

To do that, github.com/juju/utils/cert would need to be updated so that you can pass x509.ExtraExtensions into NewLeaf. Let's just leave a TODO for now, and tidy that up later.